README.md
# IOQueue

Helps you set up multiple input and output queues and link them into a
job graph. As events flow through the queues jobs will be triggered
and automatically read and write to input and output queues.

## Installation

You can install _IOQueue_ via [pip] from [PyPI]:

```console
$ pip install job_queue
```

## Usage

```python
from job_queue import JobQueue

jq  = JobQueue()


@jq.link(input_q_name="urls", output_q_name="links", batch_size=64)
def crawler(items, **env):
    links = []
    for item in items:
        url = item['url']
        links.append({'link': f"{url}/a.html", **item})
        links.append({'link': f"{url}/b.html", **item})
    return links

@jq.link(input_q_name="links", output_q_name="vecs", batch_size=128)
def transform(items, **cfg):
    vectors = []
    for item in items:
        vectors.append({'vector': [1, 2, 3], **item})
    return vectors

# Load up the DAG with initial data
urls = [dict(url=f"{idx}.com") for idx in range(256)]
jq.links['crawler'].set_inputs(urls)

# Run until all links report there's no more data
# left to process
jq.run_until_complete()

outputs = jq.links['transform'].get_outputs(1000)
print(outputs)
```

## License

Distributed under the terms of the [MIT license][license],
_IOQueue_ is free and open source software.


HISTORY.rst
=======
History
=======

0.1.0 (2022-09-08)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for job_queue."""


tests/test_job_queue.py
#!/usr/bin/env python

"""Tests for `job_queue` package."""

import pytest


from job_queue import job_queue


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


job_queue/io_queues.py
"""
IOQueues is an object that tracks the progress of items in input and output
queue and offers a simple way to track jobs finished, failed or complete
as they get places into the output queue.

"""
from mimetypes import init
import os
from uuid import uuid4

try:
    from .sqliteack_queue import AckStatus
    from .sqliteack_queue import SQLiteAckQueue
except ImportError:
    from sqliteack_queue import AckStatus
    from sqliteack_queue import SQLiteAckQueue

class IOQueues:
    _SQL_SIZE_DELTA = """
    SELECT COUNT(*) 
    FROM {input_q_name} 
    LEFT JOIN {output_q_name} 
    ON {input_q_name}.{input_q_id_column}={output_q_name}.{output_q_id_column}
    WHERE {output_q_name}.{output_q_id_column} IS NULL
     AND {input_q_name}.status < %s
    """ % (AckStatus.unack)

    _SQL_GETS = """
    SELECT {input_q_name}.*
    FROM {input_q_name} 
    LEFT JOIN {output_q_name} 
    ON {input_q_name}.{input_q_id_column}={output_q_name}.{output_q_id_column}
    WHERE {output_q_name}.{output_q_id_column} IS NULL
     AND {input_q_name}.status < %s
     LIMIT {batch_size}
    """ % (AckStatus.unack)
    initialized = False

    def __init__(self, filename, input_q_name=None, output_q_name=None, 
                 input_q_id_column=None,
                 output_q_id_column=None,
                 batch_size=1, name=None,
                 queue_kwargs={}):
        self.filename = filename
        self.input_q_name = input_q_name
        self.output_q_name = output_q_name
        self.input_q = SQLiteAckQueue(filename, table_name=input_q_name, **queue_kwargs) if input_q_name else None
        self.output_q = SQLiteAckQueue(filename, table_name=output_q_name, **queue_kwargs) if output_q_name else None
        self.batch_size = batch_size 
        self.input_q_id_column = input_q_id_column or "_id"
        self.output_q_id_column = output_q_id_column or "_id"
        self.name = name
    
    def initialize(self):
        if self.initialized:
            return 
        if self.input_q is not None:
            self.input_q.initialize()
        if self.output_q is not None:
            self.output_q.initialize()
        self.initialized = True
    
    def acks(self, keys):
        self.initialize()
        self.input_q.acks(keys)
    
    def load(self, rows):
        """ Place data rows in the output queue.
        """
        self.initialize()
        self.input_q.puts(rows)

    def puts(self, rows):
        """ Place data rows in the output queue.
        """
        self.initialize()
        self.output_q.puts(rows)

    def gets(self, batch_size=None, return_keys=False):
        """ Get an interator over batches from the input queue
        that are not in the output q.
        """
        self.initialize()
        if batch_size is None:
            batch_size = self.batch_size
        if self.input_q is None:
            return None
        query = self._SQL_GETS.format(
            input_q_name=self.input_q_name,
            output_q_name=self.output_q_name,
            input_q_id_column=self.input_q_id_column,
            output_q_id_column=self.output_q_id_column,
            batch_size=batch_size
        )
        cursor = self.input_q.con.execute(query)
        rows = cursor.fetchall()
        keys = [row[0] for row in rows]
        columns = list(map(lambda x: x[0], cursor.description))
        items = [{c: v for (c, v) in zip(columns, row) if c not in ['_id', 'timestamp', 'status']}
                 for row in rows]
        items = self.input_q.unflatten_array_columns(items)
        self.input_q.updates(keys, AckStatus.unack)
        if return_keys:
            return keys, items
        else:
            return items

    def size_ready(self):
        """ Estimate how many rows are in input q that are not 
        in the output q and also not in submitted and ongoing jobs.
        """
        self.initialize()
        if self.input_q is None:
            return 0
        query = self._SQL_SIZE_DELTA.format(
            input_q_name=self.input_q_name,
            output_q_name=self.output_q_name,
            input_q_id_column=self.input_q_id_column,
            output_q_id_column=self.output_q_id_column
            )
        cursor = self.input_q.con.execute(query)
        (n,) = cursor.fetchone()
        return n


def test_ioq_puts(n=25):
    fn = 'test_cache'
    if os.path.exists(fn):
        os.remove(fn)
    ioq = IOQueues("./test_cache", output_q_name="test_outputq")
    assert ioq.size_ready() == 0

    rows = [dict(idx=idx) for idx in range(n)]
    ioq.puts(rows)
    assert ioq.output_q.count() == n
    os.remove('./test_cache')


def test_ioq_gets(n=25):
    fn = 'test_cache'
    if os.path.exists(fn):
        os.remove(fn)
    ioq = IOQueues("./test_cache", input_q_name="test_inputq", output_q_name="test_outputq")
    assert ioq.size_ready() == 0

    rows = [dict(idx=idx) for idx in range(n)]
    ioq.load(rows)
    assert ioq.size_ready() == n

    assert len(ioq.gets(1)) == 1
    assert len(ioq.gets(2)) == 2
    assert len(ioq.gets(5)) == 5
    os.remove('./test_cache')


def test_ioq_end_to_end(n=25):
    import time
    fn = 'test_cache'
    if os.path.exists(fn):
        os.remove(fn)

    rows = [dict(idx=idx) for idx in range(n)]
    ioq0 = IOQueues("./test_cache", output_q_name="test_inputq")
    ioq0.puts(rows)

    ioq = IOQueues("./test_cache", 
                  input_q_name="test_inputq", 
                  output_q_name="test_outputq",
                  queue_kwargs=dict(timeout=0.2))
    assert ioq.size_ready() == n

    batch = ioq.gets(n * 2)
    assert len(batch) == n
    assert ioq.size_ready() == 0

    # Items recently got are unavailable
    batch2 = ioq.gets(n * 2)
    assert len(batch2) == 0
    assert ioq.size_ready() == 0

    # Now we timeout and return unack messages to the queue
    time.sleep(0.3)

    # Unack messages return to the queue
    assert ioq.size_ready() == n
    keys, batch3 = ioq.gets(n * 2, return_keys=True)
    assert len(batch3) == n

    # Ack the messages (now messages default to staying on the queue)
    ioq.acks(keys)
    assert ioq.size_ready() == 0
    os.remove(fn)


def test_ioq_e2e_join(n=25):
    fn = 'test_cache'
    if os.path.exists(fn):
        os.remove(fn)

    import time
    rows = [dict(idx=idx) for idx in range(n)]
    ioq0 = IOQueues("./test_cache", output_q_name="test_inputq")
    ioq0.puts(rows)

    # Set timeout to 0, which will effectively turn unack into ready
    ioq = IOQueues("./test_cache", 
                  input_q_name="test_inputq", 
                  output_q_name="test_outputq",
                  queue_kwargs=dict(timeout=0.0001))

    # Get messages, but with 0 timeout messages become reavailable
    batch = ioq.gets(n * 2)
    time.sleep(0.1)
    delta = ioq.size_ready()
    assert  delta == n

    # Now we process the messages from outputq -- now messages are unavailable
    transformed = [{**row, **{'processed': True}} for row in batch]
    ioq.puts(transformed)
    assert ioq.size_ready() == 0

    os.remove('./test_cache')


if __name__ == '__main__':
    test_ioq_puts()
    test_ioq_gets()
    test_ioq_end_to_end()
    test_ioq_e2e_join()

job_queue/job_queue.py
import math
import sqlite3
import os.path
from dataclasses import dataclass
from typing import Callable
from uuid import uuid4

from loguru import logger

try:
    from .io_queues import IOQueues
    from .sqliteack_queue import AckStatus
    from .sqliteack_queue import SQLiteAckQueue
except ImportError:
    from io_queues import IOQueues
    from sqliteack_queue import AckStatus
    from sqliteack_queue import SQLiteAckQueue


def submit_func_default(func, task_id, **kwargs):
    return func(task_id, **kwargs)


@dataclass
class Link:
    ioqueues: IOQueues
    function: Callable
    tasks: SQLiteAckQueue

    def set_inputs(self, rows):
        return self.ioqueues.input_q.puts(rows)
    
    def get_outputs(self, n):
        return self.ioqueues.output_q.gets(n)


class JobQueue:
    links = {}
    _task_count = {}

    def __init__(self, fn_q="queues.db", fn_tasks="tasks.db", submit_func=submit_func_default, verbose=True):
        self.fn_q = fn_q
        self.fn_tasks = fn_tasks
        self.submit_func = submit_func
        self.verbose = verbose
        if verbose:
            logger.info(f"Running on SQLite version {sqlite3.sqlite_version}")

    def link(self, taskq_kwargs={}, **kwargs):
        def wrapper(inner_func):
            name = kwargs.get('name', getattr(inner_func, '__name__', uuid4()))
            queue_kwargs = kwargs.get('queue_kwargs', {})
            queue_kwargs['verbose'] = queue_kwargs.get('verbose', self.verbose)
            q = IOQueues(self.fn_q, **kwargs)
            tasks = SQLiteAckQueue(self.fn_tasks, table_name=f"tasks_{name}", **taskq_kwargs)

            def func(task_id, **kwargs):
                tasks.acks([task_id])
                # If there's not input queue, just run the function
                # with no arguments
                args = [q.gets()] if q.input_q else []
                out_rows = inner_func(*args, **kwargs)
                if q.output_q_name:
                    q.puts(out_rows)
                # Mark this task as complete
                tasks.acks([task_id], status=AckStatus.ack_done)

            self.links[name] = Link(q, func, tasks)
            return func
        return wrapper

    def run_once(self):
        for name, link in self.links.items():
            delta = link.ioqueues.size_ready()
            n_tasks_required = int(math.ceil(delta / link.ioqueues.batch_size))
            n_tasks_active = link.tasks.active()
            while n_tasks_required >= n_tasks_active:
                self.create_task(name, link)
                n_tasks_active += 1
    
    def run_until_complete(self, **kwargs):
        while not self._check_complete():
            self.run_once(**kwargs)

    def _check_complete(self):
        completes = {}
        for name, link in self.links.items():
            completes[name] = link.ioqueues.size_ready() == 0
        return all(completes.values())

    def create_task(self, name, link):
        task_id = self._task_count.get(name, 0)
        task_cfg = {'task_index': task_id}
        logger.info(f"Creating task {task_id} for {name}")
        key = link.tasks.put(task_cfg)
        self.submit_func(link.function, key, **task_cfg)
        self._task_count[name] = task_id + 1


def test_ioq_simple(n=25):
    for batch_size in [1, 7]:
        fn = 'tasks.db'
        if os.path.exists(fn):
            os.remove(fn)
        fn = 'queues.db'
        if os.path.exists(fn):
            os.remove(fn)

        l = JobQueue(fn)

        @l.link(input_q_name="inq", output_q_name="outq", batch_size=batch_size)
        def transform(items, **cfg):
            logger.info(f"Processing {len(items)} items {cfg}")
            idxs = [{'out': item['idx'] + 50, **item} for item in items]
            return idxs

        # Load up the DAG with initial data
        inputs = [dict(idx=idx) for idx in range(n)]
        l.links['transform'].set_inputs(inputs)
        
        # Run until all links report there's no more data
        # left to process
        l.run_until_complete()

        # Check outputs are what we expected
        items = l.links['transform'].get_outputs(100)
        flat = [item['out'] for item in items]
        assert all(k in flat for k in range(50, 75))
        os.remove("tasks.db")
        os.remove("queues.db")


def test_ioq_complex(n=5, batch_size=10):
    fn = 'tasks.db'
    if os.path.exists(fn):
        os.remove(fn)
    fn = 'queues.db'
    if os.path.exists(fn):
        os.remove(fn)

    l = JobQueue(fn)

    @l.link(input_q_name="urls", output_q_name="links", batch_size=batch_size)
    def crawler(items, **cfg):
        logger.info(f"Processing {len(urls)} items {cfg} ")
        links = []
        for item in items:
            url = item['url']
            links.append({'link': f"{url}/a.html", **item})
            links.append({'link': f"{url}/b.html", **item})
        return links

    @l.link(input_q_name="links", output_q_name="vecs", batch_size=batch_size)
    def transform(items, **cfg):
        logger.info(f"Processing {len(items)} links {cfg}")
        vectors = []
        for item in items:
            vectors.append({'vector': [1, 2, 3], **item})
        return vectors

    @l.link(input_q_name="vecs", output_q_name="mean_vec", batch_size=10000)
    def sum_vector(items, **cfg):
        import numpy as np
        vecs = [item['vector'] for item in items]
        logger.info(f"Processing {len(vecs)} vecs {cfg}")
        if len(vecs) == 0:
            return []
        sum = float(np.concatenate(vecs).sum())
        rows = [dict(sum_vector=sum)]
        return rows

    # Load up the DAG with initial data
    urls = [dict(url=f"{idx}.com") for idx in range(n)]
    l.links['crawler'].set_inputs(urls)
    
    # Run until all links report there's no more data
    # left to process
    l.run_until_complete()

    # Check outputs are what we expected
    row, = l.links['sum_vector'].get_outputs(100)
    assert row['sum_vector'] == 60.0
    os.remove("tasks.db")
    os.remove("queues.db")

if __name__ == '__main__':
    test_ioq_simple()
    test_ioq_complex()

examples/example.py
from job_queue import JobQueue

jq  = JobQueue()


@jq.link(input_q_name="urls", output_q_name="links", batch_size=64)
def crawler(items, **env):
    links = []
    for item in items:
        url = item['url']
        links.append({'link': f"{url}/a.html", **item})
        links.append({'link': f"{url}/b.html", **item})
    return links

@jq.link(input_q_name="links", output_q_name="vecs", batch_size=128)
def transform(items, **cfg):
    vectors = []
    for item in items:
        vectors.append({'vector': [1, 2, 3], **item})
    return vectors

# Load up the DAG with initial data
urls = [dict(url=f"{idx}.com") for idx in range(256)]
jq.links['crawler'].set_inputs(urls)

# Run until all links report there's no more data
# left to process
jq.run_until_complete()

outputs = jq.links['transform'].get_outputs(1000)
print(outputs)